### PR TITLE
Improve device controls responsiveness and WebSocket connection resilience

### PR DESCRIPTION
**Pull request by bobaoapae**
## Summary

I've been experiencing slow device controls loading, especially when away from my home network. Opening the controls panel would show "Loading..." for several seconds before any data appeared, and certain scenarios (rapid open/close cycles, or opening the panel after force-stop) would cause controls to get stuck permanently. These are problems that have been reported by others as well (#3396, #5195, #5952, #5472).

After investigating the root causes, I identified several areas where the controls experience could be improved:

1. **Registry data (area/device/entity) is fetched fresh via WebSocket every time the controls panel opens**, even though this data rarely changes. Added an in-memory cache with a 5-minute TTL that also serves stale data when the WebSocket is disconnected, so controls don't have to wait for a reconnection to display.

2. **Entity states are only available while the controls panel is open.** Moved entity state subscription to `WebsocketManager` so states are continuously synced in background. When the panel opens, cached states are sent synchronously in `request()` — controls appear instantly without any "Loading..." state on subsequent opens. On a cold start with no in-memory cache (e.g. right after force-stop), placeholder controls with `STATUS_UNKNOWN` are emitted immediately so the panel never shows an infinite loading spinner; the async path replaces them with real data a moment later.

3. **Camera thumbnails were downloaded synchronously in `createControl()` via `runBlocking`**, blocking delivery of every control below the camera in the list and — worse — tripping `StrictMode` + `CrashFailFastHandler` when `provideControlFeatures` runs on the main thread via the synchronous cached-control path. Removed the `runBlocking` fallback entirely. Thumbnails are populated by a dedicated cache with interval-based background refresh (10s on local network, 10min on external) and force-refresh on panel open. Cache misses render the placeholder icon until the background prefetch completes.

4. **When one entity changes state (e.g. a cover moving 1% at a time), only that entity was re-sent to the subscriber.** Android's `ControlsProviderService` can mark other controls as stale when they stop being delivered. Changed to re-send all entities on any state change.

5. **WebSocket dead connection detection relies on a manual 30s ping cycle in `WebsocketManager`.** Added a dedicated `OkHttpClient` for WebSocket connections with `pingInterval(15s)` and `connectTimeout(5s)` for faster detection and failover. Also reduced `DELAY_BEFORE_RECONNECT` from 10s to 1s so the WebSocket recovers quickly from transient failures.

6. **WiFi signal strength is not considered when determining if the device is on the home network.** Added `getWifiSignalStrength()` to `WifiHelper` and a check in `isInternal()` — when WiFi RSSI drops below -80 dBm, the app proactively switches to the external/cloud URL before the connection becomes unusable.

7. **Fix a race in `WebSocketCoreImpl.createSubscriptionFlow` that caused `subscribe_entities` initial state events to be silently dropped.** The previous implementation used ``callbackFlow { ... }.shareIn(WhileSubscribed, replay = 0)``, which only starts the upstream on first subscriber. The initial state event arrived between `sendMessage` returning and the caller's `.collect` — with no subscriber yet and `replay = 0`, the event was lost and `.collect` waited forever. This reproducibly left device controls stuck on placeholders after the app process was killed. Replaced with an explicit `MutableSharedFlow(replay = 1, extraBufferCapacity = 64)` plus a relay job started *before* `sendMessage`, so the initial event is buffered in the replay cache and delivered to late subscribers. Registry update subscriptions also keep a long `SUBSCRIPTION_KEEPALIVE = 30.minutes` so background sync and subsequent panel opens reuse the same subscription without re-subscribing on every cycle.

8. **Added a generic `IgnoreSystemDiskIo` StrictMode rule** that filters `DiskReadViolation` and `DiskWriteViolation` whose stack trace contains no application frames. This replaces an earlier OEM-specific rule and covers Oppo/OnePlus (`OplusUIFirstManager`, `OplusHansManager`, `OplusBinderProxy`), Samsung, MIUI, and any future OEM disk I/O happening in binder transactions outside application control.

9. **`WebSocketCoreImpl.onMessage` logged the entire raw payload via `Timber.d`**, and in debug builds ``sensitive()`` returns the text unchanged. With coalesced registry responses reaching several MB, logcat splits the call into thousands of 4 KB chunks and the WebSocket thread blocks on logging for tens of seconds — which by itself caused post-force-stop panel opens to appear hung. Truncated the preview to 200 characters (keeping the size prefix) so large payloads log in under a millisecond and the message dispatcher doesn't starve.

After running the previous nine changes in production for a few weeks, two further problems surfaced — both share the same diagnostic insight: **WebSocket health is the right signal to gate several decisions that today rely only on local heuristics.**

10. **Cached controls were rendered as live as long as `WebSocketState == Active`.** OkHttp only transitions the socket to `Closed` after a ping/read timeout (up to ~45s after the physical link drops). During that window, panels could display values that were hours old as if they were current — a user could tap a "light off" tile from the cache and believe the command succeeded when nothing actually got sent. Added `isLiveConnectionTrustworthy(WebSocketState, hasActiveNetwork)` in `common/.../websocket/WebSocketHealth.kt`. The cache is now only displayed as live when both the WebSocket reports `Active` AND `ConnectivityManager` reports an active network; otherwise `sendCachedControlsImmediately` emits `STATUS_UNKNOWN` placeholders that the async path replaces once the real subscription delivers events. To shorten the OkHttp detection window further, `WebSocketCoreImpl` now subscribes to `NetworkChangeObserver` and force-closes the live connection the instant `ConnectivityManager` reports no network — short-circuiting the ~45s ping timeout and immediately unblocking any in-flight `connect()` call that would otherwise wait out the 30s auth timeout when the user re-enables the network.

11. **Opening the app sometimes shows the Frontend connection-error screen even when the background WebSocket is perfectly healthy.** The user could see device controls work fine in the system controls panel while the app greeted them with "Failed to connect" and a Retry button on cold start. Likely causes for the WebView's failure path: the 10s `CONNECTION_TIMEOUT` firing on slow page loads, briefly expired tokens caught mid-load, or momentary SSL glitches in Chromium. When `FrontendViewModel.onError` receives an `UnreachableError`, the synchronous gate (network available + retry budget) plus a strict re-check at fire time (`WebSocketState.Active`) now decides whether to schedule a silent retry instead of transitioning to the `Error` state. The view stays on `LoadingScreen` for the duration of the backoff window, so the error screen never flashes when the retry succeeds. Up to 3 attempts with exponential backoff (1s → 2s → 4s, capped at 5s) plus 0..500ms jitter; the strict gate is re-evaluated after the delay so a WebSocket that flips to `Closed` mid-backoff bails to the error state instead of looping. `AuthenticationError`, `UnknownError`, and `UnrecoverableError` do not auto-retry — surfacing them immediately gives the user faster feedback. `onRetry`, `switchServer`, and the `Connected → Content` transition all reset the budget and cancel any pending retry job; the retry job lives in `viewModelScope` so `onCleared` cancels it automatically.

### Related issues
- #3396 — Android Mobile App Intermittently Slow (99 comments, open)
- #5195 — Device controls takes time to load
- #5952 — Quick Settings sometimes Very Slow (~5-10s)
- #5472 — Device controls sometimes get stuck on loading
- #5259 — Gracefully handle reconnection failure on WebSocket

## Checklist
- [x] New or updated tests have been added following the testing guidelines.
- [x] The code follows the project's code style and best practices.
- [x] The changes have been thoroughly tested; edge cases considered.
- [x] Changes are backward compatible whenever feasible.

## Screenshots
N/A — no user-facing UI changes, only behavioral improvements.

## Link to pull request in documentation repositories
N/A

## Any other notes
- The `pingInterval(15s)` on the WebSocket client may have battery implications. It's configured only on the WebSocket-specific `OkHttpClient` (not the shared REST client) and only generates traffic when the WebSocket is already connected. Happy to discuss tradeoffs or make the interval configurable.
- `SUBSCRIPTION_KEEPALIVE = 30.minutes` is long enough that brief panel close/reopen cycles (and the background entity sync in `WebsocketManager`) reuse the same subscription, but short enough that the server does unsubscribe on real idle. This replaced an earlier `Duration.INFINITE` that reviewers flagged as a potential leak.
- The new `isLiveConnectionTrustworthy` helper is shared between two callers (`HaControlsProviderService` for #10 and `FrontendViewModel` for #11) and has a dedicated unit test (`IsLiveConnectionTrustworthyTest`). The two callers consume the same per-server `WebSocketRepository` singleton via `ServerManager`, so a single source of truth gates both.
- Tested on OnePlus 15 (Android 16) on both local network and 4G/cloud connections with the persistent WebSocket setting enabled.

### CODE REVIEW

File: common/src/test/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketRepositoryImplTest.kt [hunk 1]
```diff
@@ -232,6 +232,58 @@ class WebSocketRepositoryImplTest {
...
+            coEvery { webSocketCore.sendMessage(match<Map<String, Any?>> { it["type"] == "config/area_registry/list" }) } returns fakeResponse
+            coEvery { webSocketCore.getConnectionState() } returns io.homeassistant.companion.android.common.data.websocket.WebSocketState.Active
+            // mapResponse will return null for our mock, but getAreaRegistry will cache null and try again
+            // Instead, let's verify the caching behavior by calling twice and checking sendMessage count
+            repository.getAreaRegistry()
+            repository.getAreaRegistry()
+
+            // Should only call sendMessage once (second call uses cache)
+            coVerify(exactly = 1) { webSocketCore.sendMessage(match<Map<String, Any?>> { it["type"] == "config/area_registry/list" }) }
+        }
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply common/src/test/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketRepositoryImplTest.kt:295]: Addressed in 948d18ed. The cache tests were already rewritten to stub `fakeAreaResponse()` / `fakeDeviceResponse()` / `fakeEntityResponse()` with non-null decoded `result` in a later commit on this branch (see lines 246-283 of the current file). Verified by running `:common:testDebugUnitTest --tests *WebSocketRepositoryImplTest*` — passes.

File: common/src/test/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketRepositoryImplTest.kt [hunk 2]
```diff
@@ -232,6 +232,58 @@ class WebSocketRepositoryImplTest {
...
+            repository.getAreaRegistry()
+
+            // Now simulate disconnection - cache should still be used even if stale
+            coEvery { webSocketCore.getConnectionState() } returns io.homeassistant.companion.android.common.data.websocket.WebSocketState.ClosedOther
+
+            repository.getAreaRegistry()
+
+            // Still only 1 WebSocket call because cache is used when disconnected
+            coVerify(exactly = 1) { webSocketCore.sendMessage(match<Map<String, Any?>> { it["type"] == "config/area_registry/list" }) }
+        }
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply common/src/test/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketRepositoryImplTest.kt:314]: Same commit/response as #3042448379: the test was rewritten on this branch to set up the active connection, populate the cache, then switch to `ClosedOther` before a second call. Passing in the current revision.

File: app/src/main/kotlin/io/homeassistant/companion/android/controls/CameraControl.kt [hunk 3]
```diff
@@ -25,18 +25,56 @@ import timber.log.Timber
...
+        }
+        lastFetchTime[entityId] = now
+        try {
+            val bitmap = BitmapFactory.decodeStream(URL(baseUrl + entityPicture).openStream())
+            if (bitmap != null) {
+                thumbnailCache[entityId] = bitmap
+            }
+        } catch (e: Exception) {
+            Timber.e(e, "Couldn't prefetch thumbnail for $entityId")
+        }
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/controls/CameraControl.kt:60]: Already has timeouts: `URL(...).openConnection().apply { connectTimeout = 5000; readTimeout = 5000 }` (CameraControl.kt, `prefetchThumbnail`). The socket is closed via `BitmapFactory.decodeStream` consuming it to completion or the surrounding `try/catch` on exception; no lingering stream. Noted that we're not going through the shared OkHttp client here — open to switching if you'd prefer.

File: app/src/main/kotlin/io/homeassistant/companion/android/controls/CameraControl.kt [hunk 4]
```diff
@@ -25,18 +25,56 @@ import timber.log.Timber
 
 @RequiresApi(Build.VERSION_CODES.S)
 object CameraControl : HaControl {
+
+    /** In-memory cache of camera thumbnails by entity ID */
+    private val thumbnailCache = mutableMapOf<String, Bitmap>()
+
+    /** Timestamp of last fetch per entity ID */
+    private val lastFetchTime = mutableMapOf<String, Long>()
+
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/controls/CameraControl.kt:30]: Already using `ConcurrentHashMap`: `private val thumbnailCache = ConcurrentHashMap<String, Bitmap>()` and `private val lastFetchTime = ConcurrentHashMap<String, Long>()` (CameraControl.kt lines 31 and 34). This was likely flagged against an earlier revision.

File: app/src/main/kotlin/io/homeassistant/companion/android/controls/CameraControl.kt [hunk 5]
```diff
@@ -25,18 +25,56 @@ import timber.log.Timber
...
-            null
+        val entityPicture = entity.attributes["entity_picture"] as? String
+        // Use cached thumbnail (always available after first prefetch)
+        var image = thumbnailCache[entity.entityId]
+        // If no cache and baseUrl available, try to fetch synchronously (first time only)
+        if (image == null && info.baseUrl != null && entityPicture?.isNotBlank() == true) {
+            image = getThumbnail(info.baseUrl + entityPicture)
+            if (image != null) {
+                thumbnailCache[entity.entityId] = image
+            }
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/controls/CameraControl.kt]: Fixed in 948d18ed. Removed the `runBlocking` + 2s `URL.openStream()` fallback entirely. `provideControlFeatures` now only reads from `thumbnailCache`; cache misses render the placeholder icon. Prefetch is driven by `WebsocketManager` background sync and `HaControlsProviderService.refreshCameraThumbnails` (both on background coroutines). Important because `provideControlFeatures` executes on the main thread via `sendCachedControlsImmediately`, and the `runBlocking` would trip StrictMode + CrashFailFastHandler.

File: app/src/main/kotlin/io/homeassistant/companion/android/controls/HaControlsProviderService.kt [hunk 6]
```diff
@@ -74,6 +74,16 @@ class HaControlsProviderService : ControlsProviderService() {
...
+        /** Cache of last known entity states, survives service recreation */
+        internal val cachedEntities = mutableMapOf<String, Entity>()
+
+        /** Last known control IDs per server, used by WebsocketManager to keep entities synced */
+        internal val lastControlEntityIds = mutableMapOf<Int, List<String>>()
+
+        /** Last resolved base URL per server, used for instant cached control rendering */
+        @Volatile
+        internal var lastBaseUrl = mutableMapOf<Int, String>()
     }
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/controls/HaControlsProviderService.kt]: Already scoped per server: `internal val cachedEntities = ConcurrentHashMap<Int, ConcurrentHashMap<String, Entity>>()` — outer key is `serverId`, inner map is entity-id-keyed within that server (HaControlsProviderService.kt companion object). All reads/writes go through `cachedEntities[serverId]`. Same for `lastBaseUrl` and `lastControlEntityIds`. This was likely flagged against an earlier revision.

File: app/src/main/kotlin/io/homeassistant/companion/android/controls/HaControlsProviderService.kt [hunk 7]
```diff
@@ -533,8 +575,77 @@ class HaControlsProviderService : ControlsProviderService() {
...
+                val entity = cachedEntities[entityId] ?: return@forEach
+                val domain = entityId.split(".")[0]
+                val haControl = domainToHaControl[domain] ?: return@forEach
+                val info = HaControlInfo(
+                    systemId = controlId,
+                    entityId = entityId,
+                    serverId = serverId,
+                    area = getAreaForEntity(entityId, serverId),
+                    baseUrl = lastBaseUrl[serverId],
+                )
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/controls/HaControlsProviderService.kt]: The fallback is preserved via `lastControlEntityIds.keys.firstOrNull()` (HaControlsProviderService.kt, `sendCachedControlsImmediately`). We can't call the suspend `serverManager.servers().firstOrNull()` here because this runs on the main thread (`RequestHandler.handleMessage`). `lastControlEntityIds` is populated on the previous sync, so legacy-ID callers that have ever been seen before get the same fallback as the async path. For a brand-new cold-start with legacy IDs and an empty cache, placeholder controls are emitted (also added in c7515602) so the panel never stalls — the async path then replaces them with real data.

File: app/src/main/kotlin/io/homeassistant/companion/android/controls/HaControlsProviderService.kt [hunk 8]
```diff
@@ -533,8 +575,77 @@ class HaControlsProviderService : ControlsProviderService() {
...
+
+    /**
+     * Send cached controls synchronously in request() so Android never shows "loading".
+     * No suspend, no coroutine dispatch — runs directly on the calling thread.
+     */
+    @RequiresApi(Build.VERSION_CODES.S)
+    private suspend fun refreshCameraThumbnails(
+        controlIds: MutableList<String>,
+        subscriber: Flow.Subscriber<in Control>,
+    ) {
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/controls/HaControlsProviderService.kt:642]: The current KDoc on `refreshCameraThumbnails` reads: "Force-refresh camera thumbnails for the given control IDs and re-send the updated controls to the subscriber with fresh images." That matches the implementation (`suspend`, called from `ioScope.launch`, force-refreshes via `CameraControl.prefetchThumbnail(..., force = true)`, then emits each refreshed camera control). This was likely flagged against an earlier revision that had different doc text.

File: app/src/main/kotlin/io/homeassistant/companion/android/websocket/WebsocketManager.kt [hunk 9]
```diff
@@ -184,10 +191,107 @@ class WebsocketManager(appContext: Context, workerParams: WorkerParameters) :
...
+                                entityPicture,
+                                isInternal,
+                            )
+                        }
+                    } catch (e: Exception) {
+                        Timber.e(e, "Failed to prefetch thumbnail for ${entity.entityId}")
+                    }
+                }
+            }
+    }
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/websocket/WebsocketManager.kt:284]: Already scoped to a single server: `val serverEntities = HaControlsProviderService.cachedEntities[serverId] ?: return` (WebsocketManager.kt, `prefetchCameraThumbnails` line 242). The `forEach` on line 267 iterates only `serverEntities.values.filter { it.domain == "camera" }`, so it never touches other servers' entities. This was likely flagged against an earlier revision.

File: app/src/main/kotlin/io/homeassistant/companion/android/websocket/WebsocketManager.kt [hunk 10]
```diff
@@ -184,10 +191,107 @@ class WebsocketManager(appContext: Context, workerParams: WorkerParameters) :
...
+        try {
+            coroutineScope {
+                val a = async { serverManager.webSocketRepository(serverId).getAreaRegistry() }
+                val d = async { serverManager.webSocketRepository(serverId).getDeviceRegistry() }
+                val e = async { serverManager.webSocketRepository(serverId).getEntityRegistry() }
+                a.await()
+                d.await()
+                e.await()
+            }
+            Timber.d("Prefetched registries for server $serverId")
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/websocket/WebsocketManager.kt:233]: Already rethrowing `CancellationException`: WebsocketManager.kt line 234-235 has `catch (e: CancellationException) { throw e }` before the generic `Exception` catch. This was likely flagged against an earlier revision.

File: common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketRepositoryImpl.kt [hunk 11]
```diff
@@ -54,13 +54,38 @@ import kotlinx.serialization.json.intOrNull
 import okhttp3.WebSocketListener
 
 private val matterTimeout = 2.minutes
+private val REGISTRY_CACHE_TTL = 5.minutes.inWholeMilliseconds
+private val SUBSCRIPTION_KEEPALIVE = kotlin.time.Duration.INFINITE
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketRepositoryImpl.kt]: The keepalive is already finite: `private val SUBSCRIPTION_KEEPALIVE = 30.minutes` (WebSocketRepositoryImpl.kt line 58). Not `Duration.INFINITE`. This was likely flagged against an earlier revision of the constant. 30 minutes is long enough to absorb brief panel closes/reopens while still letting the subscription tear down on real idle, and that lines up with `shareIn(... WhileSubscribed(30min))`.

File: app/src/test/kotlin/io/homeassistant/companion/android/controls/CameraControlTest.kt [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+            CameraControl.prefetchThumbnail(
+                entityId = "camera.test",
+                baseUrl = "http://localhost:8123",
+                entityPicture = "/api/camera_proxy/camera.test?token=abc",
+                isInternal = true,
+                force = true,
+            )
+            // Second call within 10s interval should not re-fetch (isInternal=true, interval=10s)
+        }
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/test/kotlin/io/homeassistant/companion/android/controls/CameraControlTest.kt]: Agreed. `CameraControlTest` as it stands only exercises the code path without asserting; I'll follow up with proper `verify(exactly = ...)` / bitmap assertions once we agree on the final shape (probably next to the refactor in 948d18ed that removed `getThumbnail(runBlocking)`, since a couple of the previously-tested paths no longer exist).

File: app/src/test/kotlin/io/homeassistant/companion/android/controls/CameraControlTest.kt [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            val mockBitmap = Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888)
+            every { BitmapFactory.decodeStream(any()) } returns mockBitmap
+
+            CameraControl.prefetchThumbnail(
+                entityId = "camera.test",
+                baseUrl = "http://localhost:8123",
+                entityPicture = "/api/camera_proxy/camera.test?token=abc",
+                isInternal = true,
+                force = true,
+            )
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/test/kotlin/io/homeassistant/companion/android/controls/CameraControlTest.kt]: Moot after 948d18ed — `provideControlFeatures` no longer calls `URL.openStream()` at all (cache-only), and `prefetchThumbnail` is not exercised from the current tests. Will add proper dependency injection for the network call in the follow-up when re-introducing prefetch coverage.

File: app/src/main/kotlin/io/homeassistant/companion/android/util/IgnoreViolationRules.kt [hunk 14]
```diff
@@ -41,7 +42,8 @@
         if (violation !is IncorrectContextUseViolation) return false
 
         return violation.stackTrace.any {
-            it.fileName?.startsWith("chromium-TrichromeWebViewGoogle") == true &&
+            (it.fileName?.startsWith("chromium-TrichromeWebViewGoogle") == true ||
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/util/IgnoreViolationRules.kt]: Fixed in 948d18ed.

[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/util/IgnoreViolationRules.kt]: Fixed in 948d18ed.

File: app/src/main/kotlin/io/homeassistant/companion/android/util/IgnoreViolationRules.kt [hunk 15]
```diff
@@ -41,7 +42,8 @@
...
+                it.fileName?.startsWith("chromium-SystemWebViewGoogle") == true) &&
```
[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/util/IgnoreViolationRules.kt]: Fixed in 948d18ed.

[the remark this answers is not in the document]
Comment (bobaoapae) [reply app/src/main/kotlin/io/homeassistant/companion/android/util/IgnoreViolationRules.kt]: Fixed in 948d18ed.

File: common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImpl.kt [hunk 16]
```diff
@@ -656,7 +694,9 @@ internal class WebSocketCoreImpl(
 
     @OptIn(DelicateCoroutinesApi::class)
     override fun onMessage(webSocket: WebSocket, text: String) {
-        Timber.d("Websocket: onMessage (text: ${sensitive(text)})")
+        // Logging huge payloads (registries can be megabytes) blocks this thread for tens of seconds
+        // because logcat splits the message into thousands of small chunks. Truncate aggressively.
+        Timber.d("Websocket: onMessage (size=${text.length} preview=${sensitive(text.take(200))})")
```
Comment (TimoPtr) [inline common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImpl.kt:699]: `fun sensitive(sensitive: String): String = sensitive.takeIf { BuildConfig.DEBUG } ?: "HIDDEN"` 

It only impacts debug builds and it is on purpose.

**Review by TimoPtr — CHANGES_REQUESTED** [review]
I won't review this PR in this state. I invite you to open a PR for each of the point you've mentioned in the description. Let's go one by one since some are going to trigger some discussion.

Thanks for the time you've spent on this changes, but it is simply not possible to review everything all together. You're touching critical parts of the application that could have huge impacts on the user we need to be extract careful with any on the changes made.

**Comment by bobaoapae** [comment]
Hello, thanks for the feedback. I will try to separate things, I don't know if I will be capable of that, a lot of things are correlated. But I will keep working on that and try send pull requests with feew charges by time. Otherwise this stay here as a entry point for someone else to take a look and split if interested. 

I'm working on that for one or two months using the app on my own device and the usage became very good. But I sure comprehend the implications, a lot of core code changed.

**Comment by TimoPtr** [comment]
Closing in favor of smaller contributions.
